RFS/src/main/java/com/rfs/common/OpenSearchClient.java

```diff
@@ -257,20 +270,51 @@ public Optional<ObjectNode> getSnapshotStatus(
         }
     }
 
-    public Mono<BulkResponse> sendBulkRequest(String indexName, String body, IRfsContexts.IRequestContext context) {
+    Retry getBulkRetryStrategy() {
+        return Retry.backoff(10, Duration.ofSeconds(2)).maxBackoff(Duration.ofSeconds(60));
+    }
+
+    public Mono<BulkResponse> sendBulkRequest(String indexName, List<BulkDocSection> docs, IRfsContexts.IRequestContext context) {
         String targetPath = indexName + "/_bulk";
+        Map<String, BulkDocSection> docsMap = docs.stream().collect(Collectors.toMap(d -> d.getDocId(), d -> d));
 
-        return client.postAsync(targetPath, body, context)
-            .map(response -> new BulkResponse(response.statusCode, response.statusText, response.headers, response.body))
-            .flatMap(resp -> {
-                if (resp.hasBadStatusCode() || resp.hasFailedOperations()) {
-                    logger.error(resp.getFailureMessage());
+        return Mono.defer(() -> {
+            log.atTrace().setMessage("Creating bulk body with document ids {}").addArgument(docsMap.keySet());
+            if (docsMap.isEmpty()) {
+                return Mono.empty();
+            }
+            var body = BulkDocSection.convertToBulkRequestBody(docsMap.values());
+            return client.postAsync(targetPath, body, context)
+                .flatMap(response -> {
+                    var resp = new BulkResponse(response.statusCode, response.statusText, response.headers, response.body);
+                    if (!resp.hasBadStatusCode() && !resp.hasFailedOperations()) {
+                        return Mono.just(resp);
+                    }
+                    // Remove all successful documents for the next bulk request attempt
+                    var successfulDocs = resp.getSuccessfulDocs();
+                    successfulDocs.forEach(docsMap::remove);
+                    log.info(
+                        "After bulk request on index '{}', {} more documents have succeed, {} remain",
+                        indexName,
+                        successfulDocs.size(),
+                        docsMap.size()
+                    );
+
+                    log.error(resp.getFailureMessage());
```

Comment:
Let's include the failure message here in the same log as above. Say my log level was error, it would be really strange to see this message and not the above log given it's not actually an issue